File: lib/hecks_specializer/meta_diagnostic_validator.rb
# lib/hecks_specializer/meta_diagnostic_validator.rb
#
# Hecks::Specializer::MetaDiagnosticValidator — Phase C PC-2.
#
# First meta-specializer that regenerates a FULL Ruby class (not just
# a thin subclass shell). Reads RubyClass + RubyMethod (+ optional
# RubyConstant) fixture rows from diagnostic_validator_meta_shape.fixtures
# and emits lib/hecks_specializer/<target>.rb byte-identical.
#
# Emission pipeline:
#   1. doc block (from RubyClass.doc_snippet, verbatim)
#   2. module nesting open (from RubyClass.module_path)
#   3. class declaration + include lines
#   4. constants (if any RubyConstant rows match class_name), sorted
#      by order, each preceded/separated by a blank line
#   5. blank line
#   6. public methods (RubyMethod rows with visibility=public) in order
#   7. blank line + "private" + blank line
#   8. private methods in order
#   9. class close
#  10. IF register_target_name non-empty: blank line + register call
#      at class-close depth (inside the innermost module)
#  11. module nesting close (ends)
#
# Method bodies come from .rb.frag snippets, read raw. The specializer
# only arranges the skeleton; bodies are the author's Ruby.
#
# Subclasses override `self.target_class_name` to pick which RubyClass
# row to emit for. Default picks the first row — kept for the pilot
# (DiagnosticValidator) which was the sole row when PC-2 landed.
#
# Scope: the base class diagnostic_validator.rb (148 LoC) and
# validator_warnings.rb (113 LoC). Design generalizes — PC-3 (driver)
# and PC-4 (fixed-point of meta_subclass itself) should reuse this
# pattern with different RubyClass rows.

module Hecks
  module Specializer
    class MetaDiagnosticValidator
      include Target

      SHAPE = REPO_ROOT.join("hecks_conception/capabilities/diagnostic_validator_meta_shape/fixtures/diagnostic_validator_meta_shape.fixtures")
      TARGET_RS = REPO_ROOT.join("lib/hecks_specializer/diagnostic_validator.rb")

      # Which RubyClass fixture row to emit for. Subclasses override
      # to pick a different row. Default (nil) picks the first row.
      def self.target_class_name
        nil
      end

      def emit
        klass = pick_class
        methods = by_aggregate("RubyMethod")
                    .select { |m| m["attrs"]["class_name"] == klass["attrs"]["name"] }
                    .sort_by { |m| m["attrs"]["order"].to_i }
        public_methods = methods.select { |m| m["attrs"]["visibility"] == "public" }
        private_methods = methods.select { |m| m["attrs"]["visibility"] == "private" }

        [
          emit_doc(klass),
          emit_module_open(klass),
          emit_class_header(klass),
          emit_constants(klass),
          emit_methods(public_methods, blank_before_first: true),
          emit_private_section(private_methods),
          emit_module_close(klass),
        ].join
      end

      private

      def pick_class
        rows = by_aggregate("RubyClass")
        name = self.class.target_class_name
        row = name ? rows.find { |r| r["attrs"]["name"] == name } : rows.first
        raise "no RubyClass row matching #{name.inspect}" unless row
        row
      end

      def emit_doc(klass)
        # Doc snippet ends with `\n`; add one more for blank-line separator
        # before the module nesting begins.
        File.read(REPO_ROOT.join(klass["attrs"]["doc_snippet"])) + "\n"
      end

      # Emit "module A\n  module B\n  ..." for each segment of module_path.
      # Empty module_path = no module nesting.
      def emit_module_open(klass)
        path = klass["attrs"]["module_path"]
        return "" if path.empty?
        segments = path.split("::")
        segments.each_with_index.map do |seg, i|
          "  " * i + "module #{seg}\n"
        end.join
      end

      def emit_module_close(klass)
        path = klass["attrs"]["module_path"]
        depth = path.empty? ? 0 : path.split("::").length
        class_end = "  " * depth + "end\n"
        register = emit_register_line(klass, depth)
        module_ends = (0...depth).to_a.reverse.map { |i| "  " * i + "end\n" }.join
        class_end + register + module_ends
      end

      # If the class registers itself as a specializer target (non-empty
      # register_target_name), emit the register call inside the innermost
      # module scope, separated by a blank line from the class close.
      def emit_register_line(klass, depth)
        name = klass["attrs"]["register_target_name"]
        return "" if name.nil? || name.empty?
        indent = "  " * depth
        "\n#{indent}register :#{name}, #{klass["attrs"]["name"]}\n"
      end

      # Emit the class line + include lines. Indented to match nested
      # module depth.
      def emit_class_header(klass)
        a = klass["attrs"]
        depth = a["module_path"].empty? ? 0 : a["module_path"].split("::").length
        indent = "  " * depth
        class_line = a["base_class"].empty? \
                       ? "#{indent}class #{a["name"]}\n"
                       : "#{indent}class #{a["name"]} < #{a["base_class"]}\n"
        mixins = a["includes"].split(",").map(&:strip).reject(&:empty?)
        mixin_lines = mixins.map { |m| "#{indent}  include #{m}\n" }.join
        class_line + mixin_lines
      end

      # Emit RubyConstant rows matching this class, sorted by order.
      # Each constant line is indented one step deeper than the class,
      # preceded by a blank line (separating from include block).
      # Empty if no constants — whole block collapses.
      def emit_constants(klass)
        constants = by_aggregate("RubyConstant")
                      .select { |c| c["attrs"]["class_name"] == klass["attrs"]["name"] }
                      .sort_by { |c| c["attrs"]["order"].to_i }
        return "" if constants.empty?
        depth = klass["attrs"]["module_path"].empty? \
                  ? 0 : klass["attrs"]["module_path"].split("::").length
        indent = "  " * (depth + 1)
        lines = constants.map do |c|
          a = c["attrs"]
          "#{indent}#{a["name"]} = #{a["value_expr"]}\n"
        end.join
        "\n#{lines}"
      end

      # Emit a run of methods with correct spacing. Each method preceded
      # by a blank line (except maybe the first — controlled by flag).
      def emit_methods(methods, blank_before_first:)
        methods.each_with_index.map do |m, i|
          lead = (i == 0 && !blank_before_first) ? "" : "\n"
          lead + emit_method(m)
        end.join
      end

      def emit_private_section(private_methods)
        return "" if private_methods.empty?
        # Blank line, "private" (indented to method depth), blank line,
        # then each private method preceded by blank.
        indent = "      " # 3 levels of 2-space = 6 spaces (Hecks::Specializer::Class)
        "\n#{indent}private\n" + emit_methods(private_methods, blank_before_first: true)
      end

      def emit_method(method)
        a = method["attrs"]
        indent = "      " # 6 spaces: module → module → class
        sig = a["signature"].empty? ? "def #{a["name"]}" : "def #{a["name"]}(#{a["signature"]})"
        body = File.read(REPO_ROOT.join(a["body_snippet"]))
        "#{indent}#{sig}\n#{body}#{indent}end\n"
      end
    end

    register :meta_diagnostic_validator, MetaDiagnosticValidator

    # Phase C PC-2 extension — second full Ruby class retirement.
    # Emits lib/hecks_specializer/validator_warnings.rb from the
    # ValidatorWarnings RubyClass row in the same shape. Adds two
    # features the base exercised for the first time:
    #   - RubyConstant rows (SHAPE, TARGET_RS at class-body top)
    #   - register_target_name ("validator_warnings")
    class MetaValidatorWarnings < MetaDiagnosticValidator
      TARGET_RS = REPO_ROOT.join("lib/hecks_specializer/validator_warnings.rb")
      def self.target_class_name
        "ValidatorWarnings"
      end
    end

    register :meta_validator_warnings, MetaValidatorWarnings
  end
end

File: hecks_life/src/specializer/meta_diagnostic_validator.rs
//! Rust port of `lib/hecks_specializer/meta_diagnostic_validator.rb`.
//!
//! Emits `lib/hecks_specializer/diagnostic_validator.rb` byte-identical
//! to the Ruby specializer's output. This is a Ruby-emitting
//! specializer (not a Rust-emitting one): method bodies come from
//! `.rb.frag` snippets read VERBATIM (no leading `//`-comment strip —
//! Ruby bodies don't carry those headers), and the emission pipeline
//! assembles a full Ruby class file from RubyClass + RubyMethod (+
//! optional RubyConstant) rows.
//!
//! Emission pipeline (mirrors the Ruby `emit`):
//!   1. doc block (from RubyClass.doc_snippet, verbatim) + trailing "\n"
//!   2. require_relative lines (from RubyClass.requires) + trailing "\n"
//!   3. module nesting open (from RubyClass.module_path)
//!   4. class header + include mixins
//!   5. constants (if any), sorted by order; preceded by blank line
//!      when the class has includes (so the block breathes off them)
//!   6. public methods, sorted by order (blank-line between each)
//!   7. blank line + "private" + blank line + private methods
//!   8. class close + optional `register` line + module closes
//!
//! Default target row is `DiagnosticValidator` — the PC-2 pilot and
//! the tracked golden. PC-4 adds `meta_diagnostic_validator` /
//! `meta_validator_warnings` subclass-style emission via a different
//! row name, but the Rust-side Phase D port only needs the default
//! behavior to drive the tracked file regeneration.
//!
//! Usage:
//!   let rb = meta_diagnostic_validator::emit(repo_root)?;
//!   print!("{}", rb);
//!
//! [antibody-exempt: hecks_life/src/specializer/meta_diagnostic_validator.rs
//!  — Phase D D3 Ruby-native specializer port]

use crate::ir::Fixture;
use crate::specializer::util;
use std::error::Error;
use std::path::Path;

const SHAPE_REL: &str = "hecks_conception/capabilities/diagnostic_validator_meta_shape/fixtures/diagnostic_validator_meta_shape.fixtures";

/// Default `RubyClass` row to emit for — matches the Ruby side's
/// `self.target_class_name` default on `MetaDiagnosticValidator`
/// itself (nil → first row). The tracked output at
/// `lib/hecks_specializer/diagnostic_validator.rb` is the first row's
/// rendering.
const DEFAULT_TARGET_CLASS: &str = "DiagnosticValidator";

pub fn emit(repo_root: &Path) -> Result<String, Box<dyn Error>> {
    let shape = repo_root.join(SHAPE_REL);
    let fixtures = util::load_fixtures(&shape)?;

    let klass = pick_class(&fixtures, DEFAULT_TARGET_CLASS)?;
    let class_name = util::attr(klass, "name");

    let methods: Vec<&Fixture> = util::by_aggregate_sorted(&fixtures, "RubyMethod", "order")
        .into_iter()
        .filter(|m| util::attr(m, "class_name") == class_name)
        .collect();
    let public_methods: Vec<&Fixture> = methods
        .iter()
        .copied()
        .filter(|m| util::attr(m, "visibility") == "public")
        .collect();
    let private_methods: Vec<&Fixture> = methods
        .iter()
        .copied()
        .filter(|m| util::attr(m, "visibility") == "private")
        .collect();

    let mut out = String::new();
    out.push_str(&emit_doc(repo_root, klass)?);
    out.push_str(&emit_requires(klass));
    out.push_str(&emit_module_open(klass));
    out.push_str(&emit_class_header(klass));
    out.push_str(&emit_constants(repo_root, &fixtures, klass)?);
    out.push_str(&emit_methods(repo_root, &public_methods, true)?);
    out.push_str(&emit_private_section(repo_root, &private_methods)?);
    out.push_str(&emit_module_close(klass));
    Ok(out)
}

fn pick_class<'a>(fixtures: &'a [Fixture], name: &str) -> Result<&'a Fixture, Box<dyn Error>> {
    let rows = util::by_aggregate(fixtures, "RubyClass");
    rows.into_iter()
        .find(|r| util::attr(r, "name") == name)
        .ok_or_else(|| format!("no RubyClass row matching {:?}", name).into())
}

fn emit_doc(repo_root: &Path, klass: &Fixture) -> Result<String, Box<dyn Error>> {
    // Doc snippet ends with `\n`; add one more for blank-line separator
    // before the module nesting begins.
    let path = repo_root.join(util::attr(klass, "doc_snippet"));
    let raw = util::read_snippet_raw(&path)?;
    Ok(format!("{raw}\n"))
}

fn emit_requires(klass: &Fixture) -> String {
    let raw = util::attr(klass, "requires");
    if raw.is_empty() {
        return String::new();
    }
    let paths: Vec<&str> = raw.split(',').map(|p| p.trim()).filter(|p| !p.is_empty()).collect();
    if paths.is_empty() {
        return String::new();
    }
    let mut out = String::new();
    for p in &paths {
        out.push_str(&format!("require_relative \"{p}\"\n"));
    }
    out.push('\n');
    out
}

fn emit_module_open(klass: &Fixture) -> String {
    let path = util::attr(klass, "module_path");
    if path.is_empty() {
        return String::new();
    }
    let mut out = String::new();
    for (i, seg) in path.split("::").enumerate() {
        let indent = "  ".repeat(i);
        out.push_str(&format!("{indent}module {seg}\n"));
    }
    out
}

fn emit_module_close(klass: &Fixture) -> String {
    let path = util::attr(klass, "module_path");
    let depth = if path.is_empty() { 0 } else { path.split("::").count() };
    let class_end = format!("{}end\n", "  ".repeat(depth));
    let register = emit_register_line(klass, depth);
    let mut module_ends = String::new();
    for i in (0..depth).rev() {
        module_ends.push_str(&format!("{}end\n", "  ".repeat(i)));
    }
    format!("{class_end}{register}{module_ends}")
}

fn emit_register_line(klass: &Fixture, depth: usize) -> String {
    let name = util::attr(klass, "register_target_name");
    if name.is_empty() {
        return String::new();
    }
    let indent = "  ".repeat(depth);
    let class_name = util::attr(klass, "name");
    format!("\n{indent}register :{name}, {class_name}\n")
}

fn emit_class_header(klass: &Fixture) -> String {
    let module_path = util::attr(klass, "module_path");
    let depth = if module_path.is_empty() { 0 } else { module_path.split("::").count() };
    let indent = "  ".repeat(depth);
    let name = util::attr(klass, "name");
    let base = util::attr(klass, "base_class");
    let class_line = if base.is_empty() {
        format!("{indent}class {name}\n")
    } else {
        format!("{indent}class {name} < {base}\n")
    };
    let mixin_lines: String = util::attr(klass, "includes")
        .split(',')
        .map(|m| m.trim())
        .filter(|m| !m.is_empty())
        .map(|m| format!("{indent}  include {m}\n"))
        .collect();
    format!("{class_line}{mixin_lines}")
}

fn emit_constants(
    repo_root: &Path,
    fixtures: &[Fixture],
    klass: &Fixture,
) -> Result<String, Box<dyn Error>> {
    let _ = repo_root;
    let class_name = util::attr(klass, "name");
    let constants: Vec<&Fixture> = util::by_aggregate_sorted(fixtures, "RubyConstant", "order")
        .into_iter()
        .filter(|c| util::attr(c, "class_name") == class_name)
        .collect();
    if constants.is_empty() {
        return Ok(String::new());
    }
    let module_path = util::attr(klass, "module_path");
    let depth = if module_path.is_empty() { 0 } else { module_path.split("::").count() };
    let indent = "  ".repeat(depth + 1);
    let mut lines = String::new();
    for c in &constants {
        let cname = util::attr(c, "name");
        let val = util::attr(c, "value_expr");
        lines.push_str(&format!("{indent}{cname} = {val}\n"));
    }
    let includes = util::attr(klass, "includes").trim();
    Ok(if includes.is_empty() {
        lines
    } else {
        format!("\n{lines}")
    })
}

fn emit_methods(
    repo_root: &Path,
    methods: &[&Fixture],
    blank_before_first: bool,
) -> Result<String, Box<dyn Error>> {
    let mut out = String::new();
    for (i, m) in methods.iter().enumerate() {
        let lead = if i == 0 && !blank_before_first { "" } else { "\n" };
        out.push_str(lead);
        out.push_str(&emit_method(repo_root, m)?);
    }
    Ok(out)
}

fn emit_private_section(
    repo_root: &Path,
    private_methods: &[&Fixture],
) -> Result<String, Box<dyn Error>> {
    if private_methods.is_empty() {
        return Ok(String::new());
    }
    // 3 levels of 2-space indent = 6 spaces (module → module → class).
    let indent = "      ";
    let body = emit_methods(repo_root, private_methods, true)?;
    Ok(format!("\n{indent}private\n{body}"))
}

fn emit_method(repo_root: &Path, method: &Fixture) -> Result<String, Box<dyn Error>> {
    let indent = "      "; // 6 spaces: module → module → class
    let prefix = if util::attr(method, "receiver") == "self" {
        "def self."
    } else {
        "def "
    };
    let name = util::attr(method, "name");
    let signature = util::attr(method, "signature");
    let sig = if signature.is_empty() {
        format!("{prefix}{name}")
    } else {
        format!("{prefix}{name}({signature})")
    };
    let body_path = repo_root.join(util::attr(method, "body_snippet"));
    let body = util::read_snippet_raw(&body_path)?;
    let doc = emit_method_doc(repo_root, method)?;
    Ok(format!("{doc}{indent}{sig}\n{body}{indent}end\n"))
}

fn emit_method_doc(repo_root: &Path, method: &Fixture) -> Result<String, Box<dyn Error>> {
    let path = util::attr(method, "doc_snippet");
    if path.is_empty() {
        return Ok(String::new());
    }
    util::read_snippet_raw(&repo_root.join(path))
}
